c/cert/src/rules/EXP43-C/DoNotPassAliasedPointerToRestrictQualifiedParam.ql

```diff
@@ -0,0 +1,150 @@
+}
+
+/**
+ * A call to a function that has a parameter with a restrict-qualified pointer type.
+ */
+class CallToFunctionWithRestrictParameters extends FunctionCall {
+  CallToFunctionWithRestrictParameters() {
+    this.getTarget() instanceof FunctionWithRestrictParameters
+  }
+
+  Expr getARestrictPtrArg() {
+    result =
+      this.getArgument(this.getTarget()
+            .(FunctionWithRestrictParameters)
+            .getARestrictPtrParam()
+            .getIndex())
+  }
+
+  Expr getAPtrArg() {
+    result = this.getAnArgument() and
+    pointerValue(result)
+  }
+
+  Expr getAPossibleSizeArg() {
+    exists(Parameter param |
+      param = this.getTarget().(FunctionWithRestrictParameters).getAParameter() and
+      param.getUnderlyingType() instanceof IntegralType and
+      // exclude __builtin_object_size
+      not result.(FunctionCall).getTarget() instanceof BuiltInFunction and
+      result = this.getArgument(param.getIndex())
+    )
+  }
+}
+
+/**
+ * A `PointsToExpr` that is an argument of a pointer-type in a `CallToFunctionWithRestrictParameters`
+ */
+class CallToFunctionWithRestrictParametersArgExpr extends Expr {
+  CallToFunctionWithRestrictParametersArgExpr() {
+    this = any(CallToFunctionWithRestrictParameters call).getAPtrArg()
+  }
+}
+
+int getStatedValue(Expr e) {
+  // `upperBound(e)` defaults to `exprMaxVal(e)` when `e` isn't analyzable. So to get a meaningful
+  // result in this case we pick the minimum value obtainable from dataflow and range analysis.
+  result =
+    upperBound(e)
+        .minimum(min(Expr source | DataFlow::localExprFlow(source, e) | source.getValue().toInt()))
+}
+
+int getPointerArithmeticOperandStatedValue(CallToFunctionWithRestrictParametersArgExpr expr) {
+  result = getStatedValue(expr.(PointerArithmeticExpr).getOperand())
+  or
+  // edge-case: &(array[index]) expressions
+  result = getStatedValue(expr.(AddressOfExpr).getOperand().(PointerArithmeticExpr).getOperand())
+  or
+  // fall-back if `expr` is not a pointer arithmetic expression
+  not expr instanceof PointerArithmeticExpr and
+  not expr.(AddressOfExpr).getOperand() instanceof PointerArithmeticExpr and
+  result = 0
+}
+
+class PointerValueToRestrictArgConfig extends DataFlow::Configuration {
+  PointerValueToRestrictArgConfig() { this = "PointerValueToRestrictArgConfig" }
+
+  override predicate isSource(DataFlow::Node source) { pointerValue(source.asExpr()) }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(CallToFunctionWithRestrictParameters call |
+      sink.asExpr() = call.getAPtrArg().getAChild*()
+    )
+  }
+}
+
+from
+  CallToFunctionWithRestrictParameters call, CallToFunctionWithRestrictParametersArgExpr arg1,
+  CallToFunctionWithRestrictParametersArgExpr arg2, int argOffset1, int argOffset2
+where
+  not isExcluded(call, Pointers3Package::doNotPassAliasedPointerToRestrictQualifiedParamQuery()) and
+  arg1 = call.getARestrictPtrArg() and
+  arg2 = call.getAPtrArg() and
+  arg1 != arg2 and
```

Comment:
I recommend using `not` over `!=`:
```suggestion
  not arg1 = arg2 and
```
`!=` has some strange semantics, see https://codeql.github.com/docs/ql-language-reference/formulas/#equality